File: tree_primer/tree_primer__vertex_height.py
# 実行時間 10ms
from collections import deque

n, r = list(map(int, input().split()))
d = [-1] * n
e = [list(map(int, input().split())) for _ in [0] * (n - 1)]
g = [[] for _ in [0] * n]
q = deque()

d[r - 1] = 0
for i in range(n - 1):
    a, b = e[i]

    a -= 1
    b -= 1
    g[a].append(b)
    g[b].append(a)
q.append(r - 1)
while 0 < len(q):
    c = q.popleft()

    for i in g[c]:
        if d[i] == -1:
            d[i] = d[c] + 1
            q.append(i)
for i in range(n - 1):
    print("A" if d[e[i][0] - 1] < d[e[i][1] - 1] else "B")

File: tree_primer/tree_primer__vertex_height.cs
// 実行時間 20ms
using System.Collections.Generic;
using static System.Console;
using static System.Linq.Enumerable;
using Tpl = System.Tuple<int, int>;

internal class Program
{
    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] d = Repeat( -1, n[0] ).ToArray();
        var g = new List<int>[n[0]];
        var q = new Queue<int>( n[0] );
        Tpl[] e = new Tpl[n[0] - 1];

        foreach( int i in Range( 0, n[0] ) ) {
            g[i] = new List<int>( n[0] );
        }
        foreach( int i in Range( 0, n[0] - 1 ) ) {
            int[] a = ReadLine().Trim().Split().Select( x => int.Parse( x ) - 1 ).ToArray();

            e[i] = new Tpl( a[0], a[1] );
            g[a[0]].Add( a[1] );
            g[a[1]].Add( a[0] );
        }
        d[n[1] - 1] = 0;
        q.Enqueue( n[1] - 1 );
        while( 0 < q.Count ) {
            int c = q.Dequeue();

            foreach( int i in g[c].Where( i => d[i] == -1 ) ) {
                d[i] = d[c] + 1;
                q.Enqueue( i );
            }
        }
        foreach( int i in Range( 0, n[0] - 1 ) ) {
            WriteLine( d[e[i].Item1] < d[e[i].Item2] ? 'A' : 'B' );
        }
        return;
    }
}
